// brighterscript/dist/types/DoubleType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DoubleType = void 0;
const DynamicType_1 = require("./DynamicType");
const FloatType_1 = require("./FloatType");
const IntegerType_1 = require("./IntegerType");
const LongIntegerType_1 = require("./LongIntegerType");
class DoubleType {
    constructor(typeText) {
        this.typeText = typeText;
    }
    isAssignableTo(targetType) {
        return (targetType instanceof DoubleType ||
            targetType instanceof DynamicType_1.DynamicType);
    }
    isConvertibleTo(targetType) {
        if (targetType instanceof DynamicType_1.DynamicType ||
            targetType instanceof IntegerType_1.IntegerType ||
            targetType instanceof FloatType_1.FloatType ||
            targetType instanceof DoubleType ||
            targetType instanceof LongIntegerType_1.LongIntegerType) {
            return true;
        }
        else {
            return false;
        }
    }
    toString() {
        var _a;
        return (_a = this.typeText) !== null && _a !== void 0 ? _a : 'double';
    }
    toTypeString() {
        return this.toString();
    }
}
exports.DoubleType = DoubleType;
